sched/wqueue/kwork_thread.c

```diff
@@ -0,0 +1,297 @@
+         start = up_critmon_gettime(); \
+         worker(arg); \
+         elapsed = up_critmon_gettime() - start; \
+         if (elapsed > CONFIG_SCHED_CRITMONITOR_MAXTIME_WQUEUE) \
+           { \
+             serr("WORKER %p execute too long %"PRIu32"\n", \
+                   worker, elapsed); \
+           } \
+       } \
+     while (0)
+#else
+#  define CALL_WORKER(worker, arg) worker(arg)
+#endif
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+#if defined(CONFIG_SCHED_HPWORK)
+/* The state of the kernel mode, high priority work queue(s). */
+
+struct hp_wqueue_s g_hpwork;
+#endif /* CONFIG_SCHED_HPWORK */
+
+#if defined(CONFIG_SCHED_LPWORK)
+/* The state of the kernel mode, low priority work queue(s). */
+
+struct lp_wqueue_s g_lpwork;
+#endif /* CONFIG_SCHED_LPWORK */
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: work_thread
+ *
+ * Description:
+ *   These are the worker threads that performs the actions placed on the
+ *   high priority work queue.
+ *
+ *   These, along with the lower priority worker thread(s) are the kernel
+ *   mode work queues (also build in the flat build).
+ *
+ *   All kernel mode worker threads are started by the OS during normal
+ *   bring up.  This entry point is referenced by OS internally and should
+ *   not be accessed by application logic.
+ *
+ * Input Parameters:
+ *   argc, argv (not used)
+ *
+ * Returned Value:
+ *   Does not return
+ *
+ ****************************************************************************/
+
+static int work_thread(int argc, FAR char *argv[])
+{
+  FAR struct kwork_wqueue_s *wqueue;
+  FAR struct work_s *work;
+  worker_t  worker;
+  irqstate_t flags;
+  FAR void *arg;
+
+  wqueue = (FAR struct kwork_wqueue_s *)
+           ((uintptr_t)strtoul(argv[1], NULL, 0));
+
+  flags = enter_critical_section();
+
+  /* Loop forever */
+
+  for (; ; )
+    {
+      /* Then process queued work.  work_process will not return until: (1)
+       * there is no further work in the work queue, and (2) semaphore is
+       * posted.
+       */
+
+      nxsem_wait_uninterruptible(&wqueue->sem);
+
+      /* And check each entry in the work queue.  Since we have disabled
+       * interrupts we know:  (1) we will not be suspended unless we do
+       * so ourselves, and (2) there will be no changes to the work queue
+       */
+
+      /* Remove the ready-to-execute work from the list */
+
+      work = (FAR struct work_s *)sq_remfirst(&wqueue->q);
+      if (work && work->worker)
+        {
+          /* Extract the work description from the entry (in case the work
+           * instance by the re-used after it has been de-queued).
```

Comment:
This comment doesn't quite make sense. Perhaps you mean something like:

"Extract the work description from the entry (in case the work instance will be re-used after it has been de-queued)."

Reply:
> This comment doesn't quite make sense. Perhaps you mean something like:
> 
> "Extract the work description from the entry (in case the work instance will be re-used after it has been de-queued)."

Done, Thank you.